### ISSUE TITLE
[FEATURE REQUEST] 1:1 Button Mapping Option

### ISSUE THREAD

**Issue by jdeath** [libretro/FreeIntv#81]
Hello. Can you add the option to have a 1:1 customized button mapping for controllers with enough buttons?
I have the retronic adapters that allow using the standard intv controllers. The adapter treats the 1-9 keypad as separate buttons, not an analog stick, so cannot use the diagonals 1,3,7,9 with freeintv. I can use all other buttons by mapping the controller buttons using your scheme (e.g. map 5 button to rt). Thank you

**Comment by msheehan79** [comment]
@jdeath if its helpful, I have a fork of this core with this feature enabled on my dev2 branch here - https://github.com/msheehan79/FreeIntv/tree/dev2.  

It defines a second controller type that can be selected in the UI, that omits the On Screen Keypad feature in lieu of having all controller keys available for remapping.

**Comment by jdeath** [comment]
@msheehan79 Awesome. I'll try it. Probably way better than my hackjob!

**Comment by anonyme-x22** [comment]
Same request

**Comment by Jazzzed** [comment]
+1 for this feature request. I would like to use an actual Intellivision controller with this emulator.

**Comment by Jazzzed** [comment]
@msheehan79 I see you have done some work with this issue, but I am a newb and don't understand how the individual files in your Dev2 branch end up in the emulator. Seems that it needs to be compiled or something. That's way beyond my skills. If there is an easy way to go about this, some direction would be appreciated.

**Comment by msheehan79** [comment]
@Jazzzed yes you would need to compile it locally as its not part of the main branch. 

I will see if I can get those changes submitted to upstream if they get approved, but in the meantime if you are on Windows x64 I uploaded my locally compiled copy of the core, if it helps. Unfortunately I don't have any other compiled flavors besides Windows.

[https://easyupload.io/q18qxt](url)

**Comment by jdeath** [comment]
@msheehan79 Just to add: your core mod is awesome. I have been compiling it locally for my own use in batocera, but upstream would be great!

**Comment by RetroJoe64** [comment]
Hi, folks. I am the lucky owner of an "Ultimate PC Interface" (https://intellivision.us/interface.php) that turns Intellivision controllers (beige Model II with DB-9 connectors) and/or Colecovision controllers into bonafide Windows game controllers. Many Intv games require "kinetic" use of the 12-key numeric keypad e.g. Major League Baseball is unplayable with FreeINTV's OSD approach. **So, without further ado, can I ask @Jazzzed  or @msheehan79 to upload the compiled core for the branch of FreeIntv that supports direct mapping to specific controller keys, please?** (there is a link upthread but it's broken). Also, is there a reason these mods can't be merged into the main FreeIntv branch? I realize I'm in the minority, but logically I am not the only one who wants the "full fat" Intellivision experience :) 
![IMG_3632](https://github.com/libretro/FreeIntv/assets/122126375/755c05fa-9cdf-4a6f-900b-848c2e7528d0)

**Comment by jdeath** [comment]
I wish it was pushed upstream as well, I am not sure if a PR was ever made. It should not interfere with people running the standard way. However, you can't ask for a compiled core without specifying a distribution. Have you tried compiling it yourself? The libretro docs have a guide to compile cores. It is pretty easy if compile on the same type of machine you have. Windows subsystem for linux will even build a linux core no problem.

**Comment by RetroJoe64** [comment]
Hey, @jdeath  thanks for the rapid response! Yes, Windoze (10) is the target platform. Compiling the core is a bit more than I bargained for (just wanna play the games, man!), but I am willing to give it a go. Who knows - this could be the start of a whole "learn how to code" chapter in my career :) I'll keep you posted on results.

**Comment by msheehan79** [comment]
as far merging upstream, I'm happy to submit it as I'd rather see the feature in-use too - I had opened PR #79 as there wasn't currently Core Options set up on the core and that is needed first - if that one gets merged I can open PR's for the other features I have in my dev branches, including this one.

**Comment by Jazzzed** [comment]
Yeah that previously submitted PR is from a year and a half ago. It seems like this project has been abandoned. I personally am looking for a Linux version with 1:1 button mapping that can run on an Android game box. Looking through those instructions for compiling for Windows though, I don’t have a clue. I don’t even know what a shell or a core is. Just someone who would like to play a game as @RetroJoe64 mentioned.

**Comment by RetroJoe64** [comment]
I guess I should confess that compiling the core is probably within my technical skillset... but barely :) As far as FreeIntv being "abandoned", how do we change that status to "loved"? :) There is still a lot of passion in the retrogaming community for the Intellivision platform (it's about as "retro" as it gets after the Atari VCS !), and the only other emu for it is "jzintv", which is really funky, has a horrible antiquated UI, and is even less well-supported / documented than FreeIntv, IMHO. In any case (again, IMHO), the days of standalone / single-system emulators is over - RetroArch is the defacto platform for creating a multi-game emulation rig with a modern, seamless UI.

**Comment by msheehan79** [comment]
@RetroJoe64 if your on Windows, I uploaded the compiled core again below, as that is what I use also so you can at least give it a try until such time as these changes would be available in the upstream core.

https://easyupload.io/ne0fz9

**Comment by jdeath** [comment]
Here is a x64 linux core: https://easyupload.io/oaftd1 (compiled under Windows Subsystem for Linux). Not sure how to cross compile for Android, but i am sure there are directions somewhere.

It works on a batocera x64 distribution. See my complicated instructions here: https://forum.batocera.org/d/7620-freeintv-using-real-intv-controller

I will keep this issue open until merged.

**Comment by RetroJoe64** [comment]
Thanks, guys - hopefully, I'll have time to give everything a go over the long week-end. P.S. Responding to some of the Batocera comments, I am currently using Launch Box on Windows as my "meta front-end" (is there a better name than that?), and despite my relative naiveté about the emulation stack, clearly stuff like this belongs in the libretto cores. Legacy emulators have way too much convoluted / pathological coupling going in, likely a result of many years of "quick and dirty" hacks (e.g. to accommodate one specific game...) that never get refactored. But, I am surprised that RetroArch's "retropad" abstraction ignores the fact that many legacy consoles had out-of-the-box numeric keypads (APF M-1000, Intellvision, Colecovision / Adam, Atari 5200), and others had optional keypads (e.g. the Star Raiders controller for the Atari 2600).

**Comment by Jazzzed** [comment]
> Here is a x64 linux core: https://easyupload.io/oaftd1 (compiled under Windows Subsystem for Linux). Not sure how to cross compile for Android, but i am sure there are directions somewhere.
> 
> It works on a batocera x64 distribution. See my complicated instructions here: https://forum.batocera.org/d/7620-freeintv-using-real-intv-controller
> 
> I will keep this issue open until merged.

Whoa. That is awesome! Thank you so much @jdeath. A .SO file is exactly what I would need. I will give it a whirl this weekend. Thanks again.

**Comment by silvrgummy** [comment]
Hi I also need the windows 64 bit version of this but the link is down can anyone help me?

**Comment by jdeath** [comment]
[freeintv_libretro_Windows_X64.zip](https://github.com/libretro/FreeIntv/files/15142099/freeintv_libretro_Windows_X64.zip)

**Comment by Starpaddler** [comment]
Hi, I’m a newb to Retroarch (and  Intellivision emulation in general). What is the process to work the branch file to allow for full mapping of the Intellivision Controller?  Is there a downside to doing this?

Also, don’t understand why retroarch/free into does not currently support full mapping for the Intellivision controller within the current version?  Think adding ability to assign a keypad # to every available ported controller buttons would go along way to making Intellivision emulation much better. Please add me to the +1 column for adding this to the released build. Thanks.

**Comment by Jazzzed** [comment]
@Starpaddler 

I could be wrong, but I believe this pull request was recently approved. Whether it has been incorporated yet, I haven’t looked. But I do think that it has to be compiled specifically for Retroarch though. The attempts in the past of compiling for Linux haven’t worked out for me even though everything looked to be correct.

**Comment by msheehan79** [comment]
The PR that got approved was to add Core Options support, so not this specific feature but it is needed in order to enable this.  I've got a few more changes from my Dev branch I am trying to submit upstream to facilitate this feature, but my next PR https://github.com/libretro/FreeIntv/pull/89 hasn't been merged yet.

**Comment by Starpaddler** [comment]
Thank you!  Sounds like this is in work. Can you please post here when an update is approved and available?  Much appreciate the support!  Ended up getting retroarch as I recently set up a Nintellivision DS and thought I may as well get the emulator installed onto my home theater to play the less keypad intensive game, then I made some maps for the others that I thought could work before learning I couldn’t currently map to them.  

Thanks again. 

![Image](https://github.com/user-attachments/assets/f8454636-976a-42c8-966b-1e9ddfb911fb)
![Image](https://github.com/user-attachments/assets/602e21ea-54ef-4849-a2b6-ee800fb52981)
![Image](https://github.com/user-attachments/assets/6b7a6eb2-b8e9-4fc9-b727-1a18ef681c9a)
![Image](https://github.com/user-attachments/assets/1ad3971d-4c03-477f-8c28-dc060e08997a)
![Image](https://github.com/user-attachments/assets/cad0107b-6214-4f3c-a621-0c4e2d2db16f)
![Image](https://github.com/user-attachments/assets/f860ef50-a020-46e4-8084-fd3e2a83157e)
![Image](https://github.com/user-attachments/assets/e128fce3-ef61-4de2-8fd8-88576f91aff4)

**Comment by Drewcifer55** [comment]
+1 for 1:1 button mapping.  It would come in quite handy on the Atari Gamestation Go with custom firmware.
